src/linear_algebra/orientation.lean

```diff
@@ -0,0 +1,379 @@
+
+/-- Equivalence of nonzero vectors, in terms of same_ray. -/
+lemma equiv_iff_same_ray [nontrivial R] (v₁ v₂ : ray_vector M) :
+  v₁ ≈ v₂ ↔ same_ray R (v₁ : M) v₂ :=
+iff.rfl
+
+variables (R M)
+
+/-- A ray (equivalence class of nonzero vectors with common positive multiples) in a module. -/
+@[nolint has_inhabited_instance]
+def module.ray [nontrivial R] := quotient (ray_vector.same_ray_setoid R M)
+
+/-- An orientation of a module, intended to be used when `ι` is a `fintype` with the same
+cardinality as a basis. -/
+abbreviation orientation [nontrivial R] := module.ray R (alternating_map R M R ι)
+
+/-- A type class fixing an orientation of a module. -/
+class module.oriented [nontrivial R] :=
+(positive_orientation : orientation R M ι)
+
+variables {M}
+
+/-- The ray given by a nonzero vector. -/
+protected def ray_of_ne_zero [nontrivial R] {v : M} (h : v ≠ 0) : module.ray R M :=
+⟦⟨v, h⟩⟧
+
+/-- The rays given by two nonzero vectors are equal if and only if those vectors
+satisfy `same_ray`. -/
+lemma ray_eq_iff [nontrivial R] {v₁ v₂ : M} (hv₁ : v₁ ≠ 0) (hv₂ : v₂ ≠ 0) :
+  ray_of_ne_zero R hv₁ = ray_of_ne_zero R hv₂ ↔ same_ray R v₁ v₂ :=
+quotient.eq
+
+variables {R}
+
+/-- The ray given by a positive multiple of a nonzero vector. -/
+@[simp] lemma ray_pos_smul [nontrivial R] {v : M} (h : v ≠ 0) {r : R} (hr : 0 < r)
+  (hrv : r • v ≠ 0) : ray_of_ne_zero R hrv = ray_of_ne_zero R h :=
+begin
+  rw ray_eq_iff,
+  exact same_ray_pos_smul_left v hr
+end
+
+namespace module.ray
+
+/-- An arbitrary `ray_vector` giving a ray. -/
+def some_ray_vector [nontrivial R] (x : module.ray R M) : ray_vector M :=
+quotient.out x
+
+/-- The ray of `some_ray_vector`. -/
+@[simp] lemma some_ray_vector_ray [nontrivial R] (x : module.ray R M) :
+  (⟦x.some_ray_vector⟧ : module.ray R M) = x :=
+begin
+  rw some_ray_vector,
+  exact quotient.out_eq _
+end
+
+/-- An arbitrary nonzero vector giving a ray. -/
+def some_vector [nontrivial R] (x : module.ray R M) : M :=
+x.some_ray_vector
+
+/-- `some_vector` is nonzero. -/
+@[simp] lemma some_vector_ne_zero [nontrivial R] (x : module.ray R M) : x.some_vector ≠ 0 :=
+x.some_ray_vector.property
+
+/-- The ray of `some_vector`. -/
+@[simp] lemma some_vector_ray [nontrivial R] (x : module.ray R M) :
+  ray_of_ne_zero R x.some_vector_ne_zero = x :=
+begin
+  convert some_ray_vector_ray x,
+  rw ray_of_ne_zero,
+  congr,
+  rw subtype.ext_iff,
+  refl
+end
+
+end module.ray
+
+end ordered_comm_semiring
+
+section ordered_comm_ring
+
+local attribute [instance] ray_vector.same_ray_setoid
+
+variables (R : Type*) [ordered_comm_ring R]
+variables {M : Type*} [add_comm_group M] [module R M]
+
+/-- If two vectors are in the same ray, so are their negations. -/
+@[simp] lemma same_ray_neg_iff (v₁ v₂ : M) : same_ray R (-v₁) (-v₂) ↔ same_ray R v₁ v₂ :=
+⟨λ ⟨r₁, r₂, hr₁, hr₂, h⟩, ⟨r₁, r₂, hr₁, hr₂, by rwa [smul_neg, smul_neg, neg_inj] at h⟩,
+ λ ⟨r₁, r₂, hr₁, hr₂, h⟩, ⟨r₁, r₂, hr₁, hr₂, by rwa [smul_neg, smul_neg, neg_inj]⟩⟩
```

Comment:
Let's have the implication separately for dot notation:
```suggestion
variables {R : Type*} [ordered_comm_ring R]
variables {M : Type*} [add_comm_group M] [module R M]

/-- If two vectors are in the same ray, so are their negations. -/
lemma same_ray.neg {v₁ v₂ : M} : same_ray R v₁ v₂ → same_ray R (-v₁) (-v₂) :=
λ ⟨r₁, r₂, hr₁, hr₂, h⟩, ⟨r₁, r₂, hr₁, hr₂, by rwa [smul_neg, smul_neg, neg_inj]⟩

/-- `same_ray.neg` as an `iff`. -/
@[simp] lemma same_ray_neg_iff (v₁ v₂ : M) : same_ray R (-v₁) (-v₂) ↔ same_ray R v₁ v₂ :=
⟨λ h, by simpa only [neg_neg] using h.neg, same_ray.neg⟩
```

Reply:
I think it would be nice to also have:
```
lemma same_ray_neg_swap (v₁ v₂ : M) : same_ray R (-v₁) v₂ ↔ same_ray R v₁ (-v₂) :=
⟨λ h, by simpa only [neg_neg] using h.neg, λ h, by simpa only [neg_neg] using h.neg⟩
```